File: js/preflight.js
#!/usr/bin/env node
'use strict';
/*
 * preflight.js — guard task START (fleet workflow). Seeded from lccjs
 * scripts/preflight.js (#1036), with the two lccjs-specific paths parameterized:
 *
 *   - scratch dir:   ~/.lccjs/  ->  <scratchDir>  (default ~/.pmtools/<repo>/)
 *   - evidence dirs: docs/logs, docs/research  ->  <evidenceDirs>
 *
 *   pmtools preflight <issue> [--scratch-dir <dir>] [--evidence-dir <dir> ...]
 *
 * 1. Stamp started_iso to <scratchDir>/preflight-<issue>.iso.
 * 2. Start-of-task reads: git status, git worktree list, gh issue view.
 * 2.5 Surface in-repo evidence (<evidenceDirs>/<M>-*) for referenced #M.
 * 3. Assert the issue is OPEN; exit 1 otherwise.
 */

const { execSync } = require('node:child_process');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const config = require('./config');
const { makeDie } = require('./sh');
const { preflightIssueGate, preflightEvidence, preflightCloseCoherence, DEFAULT_EVIDENCE_DIRS } = require('./preflight_core');

function sh(cmd) {
  try {
    return execSync(cmd, { encoding: 'utf8', stdio: ['ignore', 'pipe', 'pipe'] });
  } catch (e) {
    return (e && (e.stdout || e.stderr)) || null;
  }
}

function out(s) { process.stdout.write(String(s).replace(/\n?$/, '\n')); }
const die = makeDie('preflight');
const indent = (s) => String(s || '').replace(/^/gm, '    ').replace(/\s+$/, '');

function listEvidenceFiles(evidenceDirs) {
  const acc = [];
  for (const d of evidenceDirs) {
    try { for (const name of fs.readdirSync(d)) acc.push(`${d}/${name}`); }
    catch (_) { /* dir absent — best-effort */ }
  }
  return acc;
}

function defaultScratchDir() {
  // Key the scratch dir off the MAIN checkout so worktrees don't scatter
  // ~/.pmtools/<worktree>/ dirs (#26).
  const root = config.mainRepoRoot();
  const repo = root ? path.basename(root) : 'repo';
  return path.join(os.homedir(), '.pmtools', repo);
}

function parseArgs(argv) {
  const a = { issue: '', scratchDir: null, evidenceDirs: [] };
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (t === '--scratch-dir') a.scratchDir = argv[++i];
    else if (t === '--evidence-dir') a.evidenceDirs.push(argv[++i]);
    else if (!a.issue) a.issue = t;
  }
  if (!a.scratchDir) a.scratchDir = defaultScratchDir();
  if (!a.evidenceDirs.length) a.evidenceDirs = DEFAULT_EVIDENCE_DIRS.slice();
  return a;
}

function main(argv) {
  const args = parseArgs(argv);
  const issue = String(args.issue || '').replace(/^#/, '');
  if (!/^\d+$/.test(issue)) die('usage: preflight <issue-number> [--scratch-dir D] [--evidence-dir D ...]', 2);

  const bar = '─'.repeat(58);
  out(bar); out(`  PREFLIGHT  ·  issue #${issue}`); out(bar);

  const stamp = (sh(`date '+%Y-%m-%dT%H:%M:%S%z'`) || '').trim();
  const scratch = path.join(args.scratchDir, `preflight-${issue}.iso`);
  try { fs.mkdirSync(args.scratchDir, { recursive: true }); fs.writeFileSync(scratch, stamp + '\n'); } catch (_) {}
  out(`  started_iso   ${stamp || '(date unavailable)'}`);
  out(`  saved to      ${scratch}`);
  out('');

  const status = sh('git status --short');
  out('  git status --short:');
  out(status && status.trim() ? indent(status) : '    (clean)');
  out('  git worktree list:');
  out(indent(sh('git worktree list')));
  out('');

  let info = null;
  const raw = sh(`gh issue view ${issue} --json number,title,state,body,comments`);
  if (raw) { try { info = JSON.parse(raw); } catch (_) {} }
  let body = '';
  let comments = [];
  if (info) {
    body = info.body || '';
    comments = Array.isArray(info.comments) ? info.comments : [];
    out(`  #${info.number} [${info.state}] ${info.title}`);
    out('  body:');
    out(indent(body && body.trim() ? body : '(no body)'));
    out(`  comments (${comments.length}):`);
    for (const c of comments) {
      const who = (c.author && c.author.login) || 'unknown';
      out(indent(`— @${who} (${c.createdAt || ''}):\n${(c.body || '').trim()}`));
    }
  } else {
    out(`  ⚠ gh issue view ${issue} unavailable (offline?) — skipping issue read.`);
  }
  out('');

  const refText = [body, ...comments.map((c) => c && c.body)].join('\n');
  const evidence = preflightEvidence(refText, listEvidenceFiles(args.evidenceDirs), args.evidenceDirs);
  out('  existing evidence — read these before writing findings:');
  if (evidence.length) { for (const p of evidence) out(`    • ${p}`); }
  else { out('    (none found for referenced tickets)'); }
  out('');

  // Config-coherence note (#63): claiming with pmtools but closing with a
  // non-pmtools close that may not parse br-/wt- branch names. Non-blocking.
  const enrich = config.loadEnrichmentConfig();
  const coherence = preflightCloseCoherence(enrich.claimCommand, enrich.closeCommand);
  if (coherence) process.stderr.write(`[preflight] note: ${coherence.warn}\n`);

  const gate = preflightIssueGate(info && info.state);
  if (gate.warn) out(`  ⚠ ${gate.warn}`);
  if (!gate.ok) die(gate.error);

  out(bar);
  out(`  PREFLIGHT OK  ·  #${issue} is OPEN  ·  started_iso stamped`);
  out(`  next: claim #${issue} under your agent identity (fleet workflow).`);
  out(bar);
}

if (require.main === module) main(process.argv.slice(2));

module.exports = { preflightIssueGate, preflightEvidence, preflightCloseCoherence, defaultScratchDir };

File: py/preflight.py
#!/usr/bin/env python3
"""preflight.py — guard task START (fleet workflow). Python twin of js/preflight.js
(seeded from lccjs scripts/preflight.js), reusing the pure preflight_core seams.

  preflight <issue> [--scratch-dir <dir>] [--evidence-dir <dir> ...]

  - scratch dir:   default ~/.pmtools/<repo>/ (<repo> = basename of git toplevel)
  - evidence dirs: default docs/logs, docs/research

Steps:
  1. Stamp started_iso to <scratchDir>/preflight-<issue>.iso.
  2. Start-of-task reads: git status, git worktree list, gh issue view.
  2.5 Surface in-repo evidence (<evidenceDirs>/<M>-*) for referenced #M.
  3. Assert the issue is OPEN; exit 1 otherwise (warn-and-proceed when gh offline).
"""
import json
import os
import re
import subprocess
import sys

import config
from preflight_core import preflight_issue_gate, preflight_evidence, DEFAULT_EVIDENCE_DIRS


def sh(cmd):
    """Run a command (list) and return combined best-effort stdout, else None."""
    try:
        out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                             text=True, check=True)
        return out.stdout
    except subprocess.CalledProcessError as e:
        return (e.stdout or e.stderr) or None
    except FileNotFoundError:
        return None


def out(s):
    sys.stdout.write(re.sub(r"\n?$", "\n", str(s)))


def die(msg):
    sys.stderr.write("[preflight] ✗ {}\n".format(msg))
    sys.exit(1)


def indent(s):
    return re.sub(r"\s+$", "", re.sub(r"(?m)^", "    ", str(s or "")))


def list_evidence_files(evidence_dirs):
    acc = []
    for d in evidence_dirs:
        try:
            for name in os.listdir(d):
                acc.append("{}/{}".format(d, name))
        except OSError:
            pass  # dir absent — best-effort
    return acc


def default_scratch_dir():
    # Key the scratch dir off the MAIN checkout so worktrees don't scatter
    # ~/.pmtools/<worktree>/ dirs (#26).
    root = config.main_repo_root()
    repo = os.path.basename(root) if root else "repo"
    return os.path.join(os.path.expanduser("~"), ".pmtools", repo)


def parse_args(argv):
    a = {"issue": "", "scratchDir": None, "evidenceDirs": []}
    i = 0
    n = len(argv)
    while i < n:
        t = argv[i]
        if t == "--scratch-dir":
            i += 1; a["scratchDir"] = argv[i] if i < n else None
        elif t == "--evidence-dir":
            i += 1
            if i < n:
                a["evidenceDirs"].append(argv[i])
        elif not a["issue"]:
            a["issue"] = t
        i += 1
    if not a["scratchDir"]:
        a["scratchDir"] = default_scratch_dir()
    if not a["evidenceDirs"]:
        a["evidenceDirs"] = list(DEFAULT_EVIDENCE_DIRS)
    return a


def main(argv):
    args = parse_args(argv)
    issue = re.sub(r"^#", "", str(args["issue"] or ""))
    if not re.match(r"^\d+$", issue):
        die("usage: preflight <issue-number> [--scratch-dir D] [--evidence-dir D ...]")

    bar = "─" * 58
    out(bar); out("  PREFLIGHT  ·  issue #{}".format(issue)); out(bar)

    stamp = (sh(["date", "+%Y-%m-%dT%H:%M:%S%z"]) or "").strip()
    scratch = os.path.join(args["scratchDir"], "preflight-{}.iso".format(issue))
    try:
        os.makedirs(args["scratchDir"], exist_ok=True)
        with open(scratch, "w", encoding="utf-8") as f:
            f.write(stamp + "\n")
    except OSError:
        pass
    out("  started_iso   {}".format(stamp or "(date unavailable)"))
    out("  saved to      {}".format(scratch))
    out("")

    status = sh(["git", "status", "--short"])
    out("  git status --short:")
    out(indent(status) if status and status.strip() else "    (clean)")
    out("  git worktree list:")
    out(indent(sh(["git", "worktree", "list"])))
    out("")

    info = None
    raw = sh(["gh", "issue", "view", issue, "--json", "number,title,state,body,comments"])
    if raw:
        try:
            info = json.loads(raw)
        except (ValueError, TypeError):
            info = None
    body = ""
    comments = []
    if info:
        body = info.get("body") or ""
        comments = info["comments"] if isinstance(info.get("comments"), list) else []
        out("  #{} [{}] {}".format(info.get("number"), info.get("state"), info.get("title")))
        out("  body:")
        out(indent(body if body and body.strip() else "(no body)"))
        out("  comments ({}):".format(len(comments)))
        for c in comments:
            who = (c.get("author") or {}).get("login") or "unknown"
            out(indent("— @{} ({}):\n{}".format(who, c.get("createdAt") or "", (c.get("body") or "").strip())))
    else:
        out("  ⚠ gh issue view {} unavailable (offline?) — skipping issue read.".format(issue))
    out("")

    ref_text = "\n".join([body] + [(c.get("body") or "") for c in comments])
    evidence = preflight_evidence(ref_text, list_evidence_files(args["evidenceDirs"]), args["evidenceDirs"])
    out("  existing evidence — read these before writing findings:")
    if evidence:
        for p in evidence:
            out("    • {}".format(p))
    else:
        out("    (none found for referenced tickets)")
    out("")

    gate = preflight_issue_gate(info.get("state") if info else None)
    if gate.get("warn"):
        out("  ⚠ {}".format(gate["warn"]))
    if not gate["ok"]:
        die(gate["error"])

    out(bar)
    out("  PREFLIGHT OK  ·  #{} is OPEN  ·  started_iso stamped".format(issue))
    out("  next: claim #{} under your agent identity (fleet workflow).".format(issue))
    out(bar)


if __name__ == "__main__":
    main(sys.argv[1:])
